### PR TITLE
enhance(sdk): tune get http file performance with buffer

### PR DESCRIPTION
**Pull request by tianweidut**
## Description
- fix unclosed issue for _make_file
- support HttpBufferedFileLike same as `S3BufferedFileLike`
- tune HTTP get file with timeout, stream and buffered.

## Modules
- [x] Client
- [x] Python-SDK

## Checklist
- [x] run code format and lint check
- [x] add unit test
- [ ] add necessary doc

### CODE REVIEW

**Review by anda-ren — APPROVED** [review]
lgtm

File: client/starwhale/core/dataset/store.py [hunk 1]
```diff
@@ -438,19 +453,151 @@ def sign_uri(self, uri: str) -> str:
...
+        if size < 0 or size > len(self):
+            size = len(self)
+
+        ret = self._buffer[self._pos : self._pos + size]
+        self._pos += size
+        return ret.tobytes()
+
+    def write(self, new_bytes: bytes) -> int:
+        _release_buffer = self._buffer
+        self._buffer = memoryview(self._buffer[self._pos :].tobytes() + new_bytes)
```
Comment (jialeicui) [inline client/starwhale/core/dataset/store.py:576]: `self._buffer[self._pos :].tobytes() + new_bytes` will allocate new memory and create another copy of the data, is this expected?

File: client/starwhale/core/dataset/store.py [hunk 2]
```diff
@@ -438,19 +453,151 @@ def sign_uri(self, uri: str) -> str:
...
+            timeout=timeout,
+        )
+
+        if not self.resp.ok:
+            self.resp.raise_for_status()
+
+        self._read_iter = self.resp.iter_content(buffer_size)
+
+    def _read_exhausted_raw(self) -> bytes:
+        return self.resp.raw.read()  # type: ignore
```
Comment (jialeicui) [inline client/starwhale/core/dataset/store.py:539]: Will it be better that `HttpBufferedFileLike` overrides the `read` function and just return self.resp.raw.read(size)?

Comment (tianweidut) [reply client/starwhale/core/dataset/store.py:539]: one read case:

```
read(32)  <-- first read 32 bytes
read(8)    <-- then read 8 bytes
read(-1)   <--  last read all remaining bytes
```
from the bellowing example, some bytes already maybe read into the buffer before , so we cannot read `self.resp.raw.read` directly.

Another we use the buffer for http stream read.

Comment (tianweidut) [reply client/starwhale/core/dataset/store.py]: updated

File: client/starwhale/core/dataset/store.py (hunk 1 above)
Comment (tianweidut) [reply client/starwhale/core/dataset/store.py]: updated,  use `bytearray` directly. However, sometimes write function needs allocate new memory.

**Review by jialeicui — APPROVED** [review]
LGTM

**Pull request merged**